torch/optim/optimizer.py

```diff
@@ -265,22 +350,24 @@ def _optimizer_step_code(self):
         pass
 
     @staticmethod
-    def profile_hook_step(func):
+    def profile_hook_step(func: Callable[_P, R]) -> Callable[_P, R]:
 
         @functools.wraps(func)
-        def wrapper(*args, **kwargs):
-            self, *_ = args
+        def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> R:
+            self = args[0]
+            assert isinstance(self, Optimizer)
```

Comment:
Again, not sure if adding asserts will mess with BC--is there a reason to add this?

Reply:
`self` is inferred as `object`

```console
$ mypy torch/optim/optimizer.py
torch/optim/optimizer.py:320:77: error: "object" has no attribute "_optimizer_step_pre_hooks"  [attr-defined]
torch/optim/optimizer.py:331:17: error: "object" has no attribute "_optimizer_step_code"  [attr-defined]
torch/optim/optimizer.py:334:40: error: "object" has no attribute "_optimizer_step_post_hooks"  [attr-defined]
```

These errors indicate to me that this code would certainly fail if it were not an `Optimizer`.

The only apparent usage in all of pytorch: https://github.com/search?q=repo%3Apytorch%2Fpytorch%20profile_hook_step&type=code

is in wrapping the `step` method of this class:

https://github.com/pytorch/pytorch/blob/cf8af57c4a53ea1f8947d90c26340d1b81ca0e19/torch/optim/optimizer.py#L317

where `args[0]` is necessarily an `Optimizer`.

Reply:
Alright, I do still believe there is slight BC risk here due to python ducktyping (e.g., a user may have an Optimizer_ class that somehow implements all these methods but does not subclass from Optimizer itself) but this may be a reasonably tiny risk.

Reply:
I will use `cast` if there is a concern for BC.